tsa/tsacmd/server.go

```diff
@@ -83,22 +83,57 @@ func (forward ForwardedTCPIP) Wait() {
 	forward.Logger.Debug("drained")
 }
 
-func (server *server) Serve(listener net.Listener) {
+func (server *server) Serve(listener net.Listener) error {
+	var tempDelay time.Duration
 	for {
-		c, err := listener.Accept()
+		conn, err := listener.Accept()
 		if err != nil {
-			if !strings.Contains(err.Error(), "use of closed network connection") {
-				server.logger.Error("failed-to-accept", err)
+			if strings.Contains(err.Error(), "use of closed network connection") {
+				// shutting down
+				return nil
 			}
 
-			return
+			if ne, ok := err.(net.Error); ok && ne.Temporary() {
+				if tempDelay == 0 {
+					tempDelay = 5 * time.Millisecond
+				} else {
+					tempDelay *= 2
+				}
+
+				if max := 1 * time.Second; tempDelay > max {
+					tempDelay = max
+				}
+
+				server.logger.Error("failed-to-accept-temporarily", err, lager.Data{
+					"delay": tempDelay.String(),
+				})
+
+				time.Sleep(tempDelay)
+				continue
+			}
+
+			server.logger.Error("failed-to-accept", err)
+
+			return err
 		}
 
 		logger := server.logger.Session("connection", lager.Data{
-			"remote": c.RemoteAddr().String(),
+			"remote": conn.RemoteAddr().String(),
 		})
 
-		go server.handshake(logger, c)
+		if tc, ok := conn.(*net.TCPConn); ok {
+			err := tc.SetKeepAlive(true)
+			if err != nil {
+				logger.Error("failed-to-set-tcp-keepalive", err)
+			}
+
+			err = tc.SetKeepAlivePeriod(15 * time.Second)
```

Comment:
Despite having looked at the other PR regarding keepalives yesterday, I got wondering about the number of probes before considering the connection idle. 

For those who might end up like me, check out @vito's commit message here https://github.com/concourse/concourse/commit/cac9c6d4435701757a96fb4f2a871cb4180d17d8 . tl;dr:  going with the default as Go doesn't expose a standard way of setting the number of probes.